exploration/0477-quoted-literals.md

```diff
@@ -0,0 +1,172 @@
+
+## Proposed Design
+
+_Describe the proposed solution. Consider syntax, formatting, errors, registry, tooling, interchange._
+
+Use the vertical line character, `|`, to delimit quoted strings. For example:
+
+> ```
+> {The Unix epoch is defined as {|Thu, 01 Jan 1970 00:00:00 GMT| :datetime}.}
+> ```
+
+```abnf
+literal       = quoted / unquoted
+quoted        = "|" *(quoted-char / quoted-escape) "|"
+quoted-char   = %x0-5B         ; omit \
+              / %x5D-7B        ; omit |
+              / %x7D-D7FF      ; omit surrogates
+              / %xE000-10FFFF
+quoted-escape = backslash ( backslash / "|" )
+```
+
+By being both uncommon in text content and uncommon as a string delimiter in other programming languages, the vertical line sidesteps the "inwards" and "outwards" problems of escaping.
+
+- [r1 GOOD] Writing `"` and `'` in literals doesn't require escaping them via `\`. This means no extra `\` that need escaping.
+- [r2 GOOD] Embedding messages in most code or containers doesn't require escaping the literal delimiters.
+- [r3 GOOD] Message don't have to be modified otherwise before embedding them.
+- [r4 FAIR] Vertical lines are not commonly used as string delimiters and thus can be harder to learn for beginners. There's prior art in a practice of using vertical lines as delimiters for inline code literals.
+- [r5 POOR] Vertical lines cannot be paired by parsers nor IDEs.
+
+## Alternatives Considered
+
+_What other solutions are available?_
+_How do they compare against the requirements?_
+_What other properties they have?_
+
+### Use quotation marks
+
+Early drafts of the syntax specification used double quotes to delimit literals. This changed in [#263](https://github.com/unicode-org/message-format-wg/issues/263#issue-1233590015), and was later proposed back in [#414](https://github.com/unicode-org/message-format-wg/pull/414).
+
+- [r1 POOR] Writing `"` and `'` in literals requires escaping them via `\`, which then needs to be escaped itself in code which uses `\` as the escape character (which is common).
+- [r2 FAIR] Embedding messages in certain programming languages and containers requires escaping the literal delimiters. Most notably, storing MF2 messages in JSON suffers from this. In many programming languages, however, alternatives to quotation marks exist, which could be used to allow unescaped quotes in messages. See [comment on #263](https://github.com/unicode-org/message-format-wg/issues/263#issuecomment-1430929542).
+- [r3 FAIR] One of the suggestions proposed to allow for both single and double quotation marks, and make them interchangeable in case one set was used by the inner content or surrounding code. This, however, requires directed modification of the message's body.
+- [r4 GOOD] Quotation marks are universally recognized as string delimiters.
+- [r5 POOR] Quotation marks cannot be paired by parsers nor IDEs.
```

Comment:
In general, IDEs commonly assume by default that double quotes`"` are paired, and handle them as pairs. Given sufficient context, this also applies to single quotes `'`.

Vim provides built-in functionality for selecting within quotes `vi'` or around them `va"`.

This behaviour is not "POOR", it's either "FAIR" or "GOOD".

Reply:
IDEs can assume quotes are paired, but will not know which direction the pair extends into. Consider: `aaa"bbb"ccc"ddd`; in vim you can use `vi"` to delete both `bbb` and `ccc`, depending on where your caret is positioned.

Let's switch this to FAIR. There is *some* support here, but not as good as for brackets. (I'll add a point about r5 in the section about using brackets.)

Reply:
Switched to FAIR and mentioned that most text editors have features to make working with quotes easier.